src/connector/src/source/kafka/enumerator/client.rs

```diff
@@ -94,36 +107,84 @@ impl SplitEnumerator for KafkaSplitEnumerator {
+                            tracing::info!(
+                                "reusing source {} shared kafka client to {} (ref count: {})",
+                                &context.info.source_id,
+                                &broker_address,
+                                item.ref_count + 1
+                            );
+                            Arc::new(Ok(SharedKafkaItem {
+                                client: item.client.clone(),
+                                ref_count: item.ref_count + 1,
+                            }))
+                        }
+                        Err(_) => arc_item.clone(),
+                    }
+                } else {
+                    let build_client: ConnectorResult<SharedKafkaItem> = async {
+                        let ctx_common = KafkaContextCommon::new(
+                            broker_rewrite_map,
+                            None,
+                            None,
+                            properties.aws_auth_props,
+                            properties.connection.is_aws_msk_iam(),
+                        )
+                        .await?;
+                        let client_ctx = RwConsumerContext::new(ctx_common);
+                        let client: BaseConsumer<RwConsumerContext> =
+                            config.create_with_context(client_ctx).await?;
+
+                        // Note that before any SASL/OAUTHBEARER broker connection can succeed the application must call
+                        // rd_kafka_oauthbearer_set_token() once – either directly or, more typically, by invoking either
+                        // rd_kafka_poll(), rd_kafka_consumer_poll(), rd_kafka_queue_poll(), etc, in order to cause retrieval
+                        // of an initial token to occur.
+                        // https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#a988395722598f63396d7a1bedb22adaf
+                        if properties.connection.is_aws_msk_iam() {
+                            #[cfg(not(madsim))]
+                            client.poll(Duration::from_secs(10)); // note: this is a blocking call
+                            #[cfg(madsim)]
+                            client.poll(Duration::from_secs(10)).await;
+                        }
+
+                        Ok(SharedKafkaItem {
+                            client: Arc::new(client),
+                            ref_count: 1,
+                        })
+                    }
+                    .await;
+
+                    Arc::new(build_client)
+                }
+            })
+            .await
+            .into_value();
+
+        let kafka_client: Arc<BaseConsumer<RwConsumerContext>> = match client_arc.as_ref() {
+            Ok(client) => client.client.clone(),
+            Err(e) => {
+                SHARED_KAFKA_CLIENT.remove(&properties.connection).await;
+                #[allow(rw::format_error)]
+                bail!("{}", e.as_report());
```

Comment:
Here is a lint exception because e is of type `&ConnectorError`, which cannot derive Clone (defined via `def_anyhow_newtype`) and can not `.to_owned`, but return error requires `ConnectorError`. Directly using `as_report` here to keep format.